hass-dash.el

```diff
@@ -143,193 +91,136 @@
-IDs with their properties.
-
-The `car' of a list is the group label while the `cdr' is a list of widget
-definitions for that group.
-
-'((\"Group Label\" . ((\"entity.id_example\" :label \"Human Readable Name\"))))
-
-See `hass-dash--create-widget' for widget properties.
+(defvar hass-dash-layout nil
+  "A list describing the root widget to show on the dashboard.
+In most cases, this will be a `group' widget.  You can then build a tree of
+arbitrary widgets to display on the dashboard.  You'll probably want to make use
+of hass widgets such as `hass-dash-toggle' or `hass-dash-group'.
 
 Full example:
 
-\(setq `hass-dash-layout'
- '((\"Group One\" . ((\"input_boolean.test_boolean\"
-                    :label \"Toggle entity\")
-                   (\"switch.bedroom_light\"
-                    :label \"Bedroom Light\")
-                   (\"input_boolean.test_boolean\"
-                    :label \"Turn off test boolean\"
-                    :service \"input_boolean.turn_off\")
-                   (\"automation.some_automation\")))
-   (\"Vacuum Group\" . ((\"vacuum.valetudo_vacuum\"
-                       :label \"Vacuum\")
-                      (\"vacuum.valetudo_vacuum\"
-                       :label \"Vacuum return home\"
-                       :service \"vacuum.return_to_base\"
-                       :state nil
-                       :icon nil)))))"
-  :group 'hass-dash
-  :type 'list)
-
-
-;; Default formatters
-(defcustom hass-dash-default-widget-formatter #'hass-dash-widget-formatter
-  "The function called to format the widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
-
-(defcustom hass-dash-default-label-formatter #'hass-dash-label-formatter
-  "The function called to format the label of widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
-
-(defcustom hass-dash-default-state-formatter #'hass-dash-state-formatter
-  "The function called to format the state of widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
-
-(defcustom hass-dash-default-icon-formatter #'hass-dash-icon-formatter
-  "The function called to format the icon of widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
+\(setq hass-dash-layout
+      \\=`(hass-dash-group
+        :title \"Home Assistant\"
+        :format \"%t\\n\\n%v\"
+        (hass-dash-group
+         :title \"Kitchen\"
+         :title-face outline-2
+         (hass-dash-toggle :entity-id \"light.kitchen_lights\")
+         (hass-dash-toggle :entity-id \"switch.entry_light\"
+                           :label \"Hallway\"
+                           :confirm t))))")
 
 
 ;; Helper functions
-(defun hass-dash--default-service-of (entity-id)
-  "Return the default service to be called for ENTITY-ID."
-  (let ((domain (hass--domain-of-entity entity-id)))
-    (cdr (assoc domain hass-dash--default-services))))
-
 (defun hass-dash--track-layout-entities ()
   "Tracks referenced entities in `hass-dash-layout' and update their state."
```

Comment:
This function doesn't seem to work for me using your layout example but with my entity IDs. Even running it manually, the `hass-tracked-entities` is still set to `nil`. I should contain a list of all entities IDs that a widget state is associated with.

Reply:
ahh, this broke as a result of adding the `hass-dash-group` widget, since I'm explicitly looking for `group` widgets to descend into children.  There's a better way to do this that uses the `:children` prop instead, I'll switch to that.